python/sgl_jax/srt/models/qwen2_moe.py

```diff
@@ -0,0 +1,605 @@
+
+        return hidden_states, residual, kv_fused, layer_callback_flag
+
+
+class Qwen2MoeModel(nnx.Module):
+    """Qwen2 MoE model with embedding, layers, and normalization."""
+
+    def __init__(
+        self,
+        config: PretrainedConfig,
+        dtype: jnp.dtype = jnp.bfloat16,
+        rngs: nnx.Rngs = None,
+        mesh: jax.sharding.Mesh = None,
+    ):
+        self.config = config
+        self.padding_idx = config.pad_token_id
+        self.vocab_size = config.vocab_size
+
+        self.embed_tokens = Embed(
+            num_embeddings=config.vocab_size,
+            features=config.hidden_size,
+            rngs=rngs,
+            dtype=dtype,
+            param_dtype=dtype,
+        )
+
+        self.layers = nnx.data(
+            [
+                Qwen2MoeDecoderLayer(
+                    config=config,
+                    layer_id=i,
+                    dtype=dtype,
+                    rngs=rngs,
+                    mesh=mesh,
+                )
+                for i in range(config.num_hidden_layers)
+            ]
+        )
+
+        self.norm = RMSNorm(
+            config.hidden_size,
+            epsilon=config.rms_norm_eps,
+            param_dtype=dtype,
+            scale_init=nnx.with_partitioning(init_fn, (None,)),
+            rngs=rngs,
+        )
+
+    def __call__(
+        self,
+        forward_batch: ForwardBatch,
+        token_to_kv_pool: KVCache,
+    ):
+        residual = None
+        hidden_states = self.embed_tokens(forward_batch.input_ids)
+        layers_kv_fused = []
+        layers_callback_flag = []
+
+        for layer in self.layers:
+            hidden_states, residual, kv_fused, callback_flag = layer(
+                forward_batch.positions,
+                hidden_states,
+                forward_batch,
+                token_to_kv_pool,
+                residual,
+            )
+            layers_kv_fused.append(kv_fused)
+            layers_callback_flag.extend(callback_flag)
+
+        if residual is not None:
+            hidden_states += residual
+        hidden_states = self.norm(hidden_states)
+
+        return hidden_states, layers_kv_fused, layers_callback_flag
+
+
+class Qwen2MoeForCausalLM(nnx.Module):
+    """Qwen2 MoE model for causal language modeling."""
+
+    def __init__(
+        self,
+        config: PretrainedConfig,
+        dtype: jnp.dtype = jnp.bfloat16,
+        rngs: nnx.Rngs = None,
+        mesh: jax.sharding.Mesh = None,
+    ):
+        self.mesh = mesh
+        self.config = config
+        self.dtype = dtype
+        logger.info("Qwen2MoeForCausalLM config dtype: %s", self.dtype)
+        self.transformer = Qwen2MoeModel(config, dtype=self.dtype, rngs=rngs, mesh=mesh)
+        if not getattr(self.config, "tie_word_embeddings", True):
+            self.lm_head = ParallelLMHead(config.vocab_size, config.hidden_size, rngs=rngs)
+        self.logits_processor = LogitsProcessor(config.vocab_size, self.mesh)
```

Comment:
The current default configuration performs TP splitting on the vocab size dimension for both embed tokens and LM heads. Please verify and modify as needed.